model/flow/sealing_segment_test.go

```diff
@@ -0,0 +1,223 @@
+
+		err := builder.AddBlock(&block1)
+		require.ErrorIs(t, err, flow.ErrSegmentResultLookup)
+	})
+
+	t.Run("should return ErrSegmentInvalidBlockHeight if block has invalid height", func(t *testing.T) {
+		block1 := unittest.BlockFixture()
+		block2 := unittest.BlockFixture()
+		block3 := unittest.BlockWithParentFixture(block2.Header)
+		resultLookup := func(flow.Identifier) (*flow.ExecutionResult, error) { return nil, nil }
+		builder := flow.NewSealingSegmentBuilder(resultLookup)
+
+		err := builder.AddBlock(&block1)
+		require.NoError(t, err)
+
+		err = builder.AddBlock(&block3)
+		require.ErrorIs(t, err, flow.ErrSegmentInvalidBlockHeight)
+	})
+}
+
+// TestSealingSegmentBuilder_SealingSegment checks behavior and sanity checks when building sealing segment
+func TestSealingSegmentBuilder_SealingSegment(t *testing.T) {
+	t.Run("should return valid root sealing segment", func(t *testing.T) {
+		resultLookup := func(flow.Identifier) (*flow.ExecutionResult, error) { return nil, nil }
+		builder := flow.NewSealingSegmentBuilder(resultLookup)
+
+		block1 := unittest.BlockFixture()
+		block1.Header.View = 0
+		err := builder.AddBlock(&block1)
+		require.NoError(t, err)
+
+		segment, err := builder.SealingSegment()
+		require.NoError(t, err)
+
+		unittest.AssertEqualBlocksLenAndOrder(t, []*flow.Block{&block1}, segment.Blocks)
+		require.Equal(t, segment.Highest().ID(), block1.ID())
+		require.Equal(t, segment.Lowest().ID(), block1.ID())
+	})
+
+	t.Run("should return valid sealing segment", func(t *testing.T) {
+		resultLookup := func(flow.Identifier) (*flow.ExecutionResult, error) { return nil, nil }
+		builder := flow.NewSealingSegmentBuilder(resultLookup)
+
+		block1 := unittest.BlockFixture()
+		block2 := unittest.BlockWithParentFixture(block1.Header)
+		receipt, seal := unittest.ReceiptAndSealForBlock(&block1)
+		block2.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receipt), unittest.WithSeals(seal)))
+
+		err := builder.AddBlock(&block1)
+		require.NoError(t, err)
+		err = builder.AddBlock(&block2)
+		require.NoError(t, err)
+
+		segment, err := builder.SealingSegment()
+		require.NoError(t, err)
+
+		require.Equal(t, segment.Highest().ID(), block2.ID())
+		require.Equal(t, segment.Lowest().ID(), block1.ID())
+	})
+
+	// test that when the highest block in the segment does not contain seals but the first ancestor of highest
+	// does contain a seal for lowest we return a valid sealing segment.
+	t.Run("should return valid sealing segment if ancestor of highest seals lowest", func(t *testing.T) {
+		resultLookup := func(flow.Identifier) (*flow.ExecutionResult, error) { return nil, nil }
+		builder := flow.NewSealingSegmentBuilder(resultLookup)
+
+		block1 := unittest.BlockFixture()
+		block2 := unittest.BlockWithParentFixture(block1.Header)
+		receipt, seal := unittest.ReceiptAndSealForBlock(&block1)
+		block2.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receipt), unittest.WithSeals(seal)))
```

Comment:
Hm, I think this is wrong. Well, this is not wrong in this test case. But this will be an invalid block since we [require](https://github.com/onflow/flow-go/blob/master/module/validation/seal_validator.go#L161) that there is at least one block between block incorporating result and block incorporating the seal. I think we should construct a valid segment.

Reply:
https://github.com/onflow/flow-go/pull/1533/commits/60a5951058ff9c214eb0b3d85afc5975b50a7721